doc/decisions/DR016-Related-entities-result-paging.md

```diff
@@ -0,0 +1,352 @@
+* Offset based designs:
+  - [Atlassian](https://developer.atlassian.com/server/framework/atlassian-sdk/atlassian-rest-api-design-guidelines-version-1/#standard-query-parameters-in-uris)
+  - [Australian Government](https://apiguide.readthedocs.io/en/latest/build_and_publish/batching_results.html)
+  - [Microsoft](https://github.com/Microsoft/api-guidelines/blob/master/Guidelines.md#982-client-driven-paging)
+* Page token based designs:
+  - [Google Cloud guidelines](https://cloud.google.com/apis/design/design_patterns#list_pagination)
+  - [Google Cloud Datastore](https://cloud.google.com/datastore/docs/concepts/queries#datastore-datastore-cursor-paging-python)
+  - [Twitter](https://developer.twitter.com/en/docs/twitter-api/pagination)
+  - [Slack](https://slack.engineering/evolving-api-pagination-at-slack/)
+* Cursor object designs:
+  - [Cisco](https://github.com/CiscoDevNet/api-design-guide#363-get)
+  - [Cloudfoundry](https://github.com/cloudfoundry/cc-api-v3-style-guide#pagination)
+  - [JDBC](https://docs.oracle.com/javase/8/docs/api/java/sql/ResultSet.html)
+  - [MongoDB](https://www.mongodb.com/docs/manual/reference/method/js-cursor/)
+  - [Python PEP 239 DB-API](https://peps.python.org/pep-0249/#fetchmany)
+    used by many database libraries.
+  - [Ruby on Rails](https://api.rubyonrails.org/v7.0.4.2/classes/ActiveRecord/Batches.html#method-i-find_in_batches)
+  - [W3C IndexedDB Cursor API spec](https://w3c.github.io/IndexedDB/#idbcursor)
+
+### Rejected options
+
+In order to avoid considering a long list of options that are either
+clearly inferior, or broadly similar to other options, some valid
+solutions were rejected early.
+
+#### Whole-batch offset
+
+The classic offset-based solution was rejected due to the critical
+drawback of unstable paging, which a page token based solution overcomes
+without introducing much more complexity.
+
+```python
+SuccessCallback = Callable[[int, List[EntityReferences]], None]
+
+def relatedEntities(
+        batch: List[EntityReference],
+        relationship,
+        successCb: SuccessCallback,
+        errorCb,
+        limit: int = -1, offset: int = 0): ...
+```
+
+##### Pros
+
+- Conceptually simple for the host.
+- Jumping arbitrarily to nth page is trivial.
+- Retains batch-awareness for each page.
+- Transactions, indexing and caching technologies exist that can
+  alleviate stability and performance problems, to some extent.
+
+##### Cons
+
+- Anomalous page results if insertions/deletions can happen before next
+  page is retrieved, or if ordering is not stable.
+- Poor performance on the back-end for large offsets into non-trivial
+  (i.e. non-indexed/cached) queries, where many rows must be iterated
+  through and skipped over.
+
+#### Whole-batch cursor object
+
+Paging through an entire batch via a cursor object was rejected mainly
+due to being broadly similar to a whole-batch page token. A whole-batch
+cursor object offers few additional benefits, whilst presenting a
```

Comment:
Im not 100% convinced that this option should be so easily discarded if we are also considering a whole-batch token. The ability of offer proper navigation of pages without new methods is quite significant I think.

Reply:
We were desperately trying to keep the number of options to a minimum. This was made difficult because we didn't think the two concerns of batch and paging strategy were orthogonal enough to consider in isolation.  

So we deliberately considered only our favoured options that sample across both the batch and paging strategy axes, accepting the possibility that the chosen solution will be some hybrid.

Back on topic... do you think this particular rejected solution is worth promoting to an "Option 4"?

Reply:
I think perhaps rephrasing as comparing options on the two axis that make up the matrix might be easier to grok?

Reply:
Split pros/cons of page tokens in 76f85a78

Also tentatively promoted this option to "Option 2.3" in c98a580e